File: src/core/fusion.rs
use super::{AgeGroup, InferenceSignal, SignalSource};

pub fn fuse(signals: Vec<InferenceSignal>) -> InferenceSignal {
    match signals.len() {
        0 => InferenceSignal::new(SignalSource::Local),
        1 => signals.into_iter().next().unwrap(),
        _ => blend(signals),
    }
}

fn blend(signals: Vec<InferenceSignal>) -> InferenceSignal {
    let mut result = InferenceSignal::new(SignalSource::Local);

    // Gender: average
    // result.gender_male = signals.iter().map(|s| s.gender_male).sum::<f64>() / count;
    // result.gender_female = signals.iter().map(|s| s.gender_female).sum::<f64>() / count;
    // result.gender_other = signals.iter().map(|s| s.gender_other).sum::<f64>() / count;

    // Gender: average only signals with data
    let gender_signals: Vec<_> = signals.iter().filter(|s| s.has_gender_signal()).collect();
    if !gender_signals.is_empty() {
        let count = gender_signals.len() as f64;
        result.gender_male = gender_signals.iter().map(|s| s.gender_male).sum::<f64>() / count;
        result.gender_female = gender_signals.iter().map(|s| s.gender_female).sum::<f64>() / count;
        result.gender_other = gender_signals.iter().map(|s| s.gender_other).sum::<f64>() / count;
    }

    // // Age: blend distributions
    // let age_dists: Vec<[f64; 7]> = signals
    //     .iter()
    //     .map(|s| {
    //         [
    //             s.age_group_under_18,
    //             s.age_group_18_24,
    //             s.age_group_25_34,
    //             s.age_group_35_44,
    //             s.age_group_45_54,
    //             s.age_group_55_64,
    //             s.age_group_65_plus,
    //         ]
    //     })
    //     .collect();
    // let weights = vec![1.0; signals.len()];
    // result.set_age_probs(AgeGroup::blend(&age_dists, &weights));

    // Age: blend only signals with data
    let age_signals: Vec<_> = signals.iter().filter(|s| s.has_age_signal()).collect();
    if !age_signals.is_empty() {
        let age_dists: Vec<[f64; 7]> = age_signals
            .iter()
            .map(|s| {
                [
                    s.age_group_under_18,
                    s.age_group_18_24,
                    s.age_group_25_34,
                    s.age_group_35_44,
                    s.age_group_45_54,
                    s.age_group_55_64,
                    s.age_group_65_plus,
                ]
            })
            .collect();
        let weights = vec![1.0; age_signals.len()];
        result.set_age_probs(AgeGroup::blend(&age_dists, &weights));
    }

    // Birth year: first non-None (local is deterministic)
    result.birth_year = signals.iter().find_map(|s| s.birth_year);

    // Organization: first non-None
    // result.organization = signals.iter().find_map(|s| s.organization.clone());

    result.organization = signals
        .iter()
        .filter_map(|s| s.organization.clone())
        .max_by_key(|org| org.name.is_some() as u8);

    // Ethnicity: highest confidence wins
    if let Some(best) = signals
        .iter()
        .filter(|s| s.ethnicity.is_some())
        .max_by(|a, b| {
            a.ethnicity_confidence
                .partial_cmp(&b.ethnicity_confidence)
                .unwrap()
        })
    {
        result.ethnicity = best.ethnicity.clone();
        result.ethnicity_confidence = best.ethnicity_confidence;
    }

    // Reasoning: combine all
    result.reasoning = signals.iter().flat_map(|s| s.reasoning.clone()).collect();

    // Latency: max (parallel execution)
    result.latency_ms = signals.iter().map(|s| s.latency_ms).max().unwrap_or(0);

    // Tokens: sum
    let total: u32 = signals.iter().filter_map(|s| s.tokens_used).sum();
    result.tokens_used = if total > 0 { Some(total) } else { None };

    result
}

File: aura-python/aura/core/fusion.py
"""Signal fusion - combines multiple agent signals into a unified result."""

from .models import InferenceSignal
from .types import SignalSource, AgeGroup


def fuse(signals: list[InferenceSignal]) -> InferenceSignal:
    """Fuse multiple inference signals into a single unified signal.

    Args:
        signals: List of signals from different agents.

    Returns:
        A single fused InferenceSignal.
    """
    if not signals:
        return InferenceSignal(source=SignalSource.LOCAL)
    if len(signals) == 1:
        return signals[0]
    return _blend(signals)


def _blend(signals: list[InferenceSignal]) -> InferenceSignal:
    """Blend multiple signals using weighted averaging and selection rules.

    Fusion strategy:
    - Gender: Average probabilities from signals with gender data
    - Age: Blend distributions from signals with age data
    - Birth year: First non-None value (deterministic)
    - Organization: Prefer one with name field populated
    - Ethnicity: Highest confidence wins
    - Reasoning: Concatenate all
    - Latency: Max (parallel execution)
    - Tokens: Sum
    """
    result = InferenceSignal(source=SignalSource.LOCAL)

    # Gender: average only signals with data
    gender_signals = [s for s in signals if s.has_gender_signal()]
    if gender_signals:
        count = len(gender_signals)
        result.gender_male = sum(s.gender_male for s in gender_signals) / count
        result.gender_female = sum(s.gender_female for s in gender_signals) / count
        result.gender_other = sum(s.gender_other for s in gender_signals) / count

    # Age: blend only signals with data
    age_signals = [s for s in signals if s.has_age_signal()]
    if age_signals:
        age_dists = [s.get_age_probs() for s in age_signals]
        weights = [1.0] * len(age_signals)
        result.set_age_probs(AgeGroup.blend(age_dists, weights))

    # Birth year: first non-None (local is deterministic)
    result.birth_year = next((s.birth_year for s in signals if s.birth_year), None)

    # Organization: prefer one with name populated
    orgs = [s.organization for s in signals if s.organization]
    if orgs:
        # Sort by whether name is present, take the one with most data
        result.organization = max(orgs, key=lambda o: o.name is not None)

    # Ethnicity: highest confidence wins
    eth_signals = [s for s in signals if s.ethnicity]
    if eth_signals:
        best = max(eth_signals, key=lambda s: s.ethnicity_confidence)
        result.ethnicity = best.ethnicity
        result.ethnicity_confidence = best.ethnicity_confidence

    # Reasoning: combine all
    result.reasoning = [r for s in signals for r in s.reasoning]

    # Latency: max (parallel execution)
    result.latency_ms = max(s.latency_ms for s in signals) if signals else 0

    # Tokens: sum
    total_tokens = sum(s.tokens_used or 0 for s in signals)
    result.tokens_used = total_tokens if total_tokens > 0 else None

    return result
